### PR TITLE
New gas work

### PR DESCRIPTION
**Pull request by mhhf**
this replaces gas rebased rebased branch

### CODE REVIEW

**Comment by livnev** [comment]
@mhhf there seems to be a bug where `klab build --exact-gas` will insert the gas expression for the `_fail` spec into the `_pass` spec.

File: libexec/klab-compile [hunk 1]
```diff
@@ -113,7 +113,9 @@ const behaviour = genBehaviour({
         ...boot
       });
 // console.log(JSON.stringify(behaviour, false, 2));
-boot.status = behaviour.success ? "SUCCESS" : "fail"
+boot.status = behaviour.success
+  ? "SUCCESS"
```
Comment (MrChico) [inline libexec/klab-compile]: replace SUCCES -> ACCEPT

File: lib/build.js [hunk 2]
```diff
@@ -339,13 +342,76 @@ const parseAct = (act_str, silent = false) => {
...
   if(!act.storage) act.storage = {};
   if(!act.storage.ACCT_ID) act.storage.ACCT_ID = []
+  const storage = {
+    ...act.storage,
+    [act.subject]: act.storage.ACCT_ID
+  }
+  delete storage.ACCT_ID;
+
+  // GAS
+  const gas = "VGas => " + ((act.hasGas && !oog) ? "VGas -Int " + act.gas : "_");
```
Comment (MrChico) [inline lib/build.js]: The post gas should not be the cleaned up gas expression, but rather the "raw" expression, just tokenized.

File: libexec/klab-compile (hunk 1 above)
Comment (livnev) [reply libexec/klab-compile]: :+1:

File: lib/build.js (hunk 2 above)
Comment (livnev) [reply lib/build.js]: :+1:

**Comment by livnev** [comment]
@mhhf Looking at the `multipleCalls` example, it seems like the file `Callee_tempDelta_pass.k` contains specs for both `Callee_add_pass` and `Callee_add_fail`. I wonder why it has included the reachability claim for the latter? It seems logically unrelated. And as a result `klab run` complains that `add_fail` needs to be proved first, which I don't think is the case.

**Comment by livnev** [comment]
There's another outstanding issue: if running the `VGas > 300000` version we discover that the final gas term is `f(VGas)`, and we solve to find an expression `G` such that `f(VGas) >= 0 iff VGas >= G`, it does **NOT** mean necessarily that `f(VGas) = VGas - G`. As an example, after going through the yoga of `klab build` `klab run` `klab solve-gas` `klab build --exact-gas`, we find that the spec `easyNest_calling_pass` contains the following gas rewrite:

```
...
        <wordStack> .WordStack => _ </wordStack>
        <localMem> .Map => _ </localMem>
        <pc> 0 => _ </pc>
        <gas> VGas => VGas -Int #if ( notBool( ABI_x ==K 0 ) ) #then 22773 #else 7535 #fi
         </gas>
        <memoryUsed> 0 => _ </memoryUsed>
        <previousGas> _ => _ </previousGas>
...
```
where the right hand side of the rewrite in the `<gas>` cell is `VGas` minus the minimum gas expression that `klab solve-gas` found, which is incorrect. It should instead read:
```
...
        <wordStack> .WordStack => _ </wordStack>
        <localMem> .Map => _ </localMem>
        <pc> 0 => _ </pc>
        <gas> VGas => ( ( ( ( VGas -Int 1257 ) -Int ( 700 +Int ( ( VGas -Int 1957 ) -Int ( ( VGas -Int 1957 ) /Int 64 ) ) ) ) +Int `#if_#then_#else_#fi_K-EQUAL`( notBool( ( ABI_x ==Int 0 )  ),( ( ( VGas -Int 1957 ) -Int ( ( VGas -Int 1957 ) /Int 64 ) ) -Int 20491 ),( ( ( VGas -Int 1957 ) -Int ( ( VGas -Int 1957 ) /Int 64 ) ) -Int 5491 ) ) ) -Int 42 )
         </gas>
        <memoryUsed> 0 => _ </memoryUsed>
        <previousGas> _ => _ </previousGas>
...
```
where the RHS is the actual raw gas expression we discovered on an earlier run.

**Comment by livnev** [comment]
Reminder of what we discussed the other day: the current preconditions for the `_pass_oog`, `_fail_oog`, and `_fail` are incorrect in specs where there is a `CALL`. If we OOG during a `CALL` then we will _usually_ `REVERT`, provided that the return value check sends us to a `REVERT` branch, like in solidity calls, etc.

It looks like the best way to implement this might be to use the new `split` construct at every `CALL` that we come across when proving a `_fail` spec and then explore the following branches:
(i) `CALL` succeeds
(i) have preconditions for success in the spec of the called method, but we OOG
(ii) have preconditions for revert in the spec of the called method, but we OOG
(iii) have preconditions for revert in the spec of the called method, and we don't OOG
They correspond to the specs `Callee_pass`, `Callee_pass_oog`, `Callee_fail_oog`, `Callee_fail`, respectively, so then those reachability rules must be imported and applied to the four branches.

Calculating the minimum gas for `_fail` is now trickier, gas needs to be such that we either have enough gas for the whole execution, or we run out of gas inside of a `CALL`. The cases where we run out of gas outside of a `CALL` are covered by `_pass_oog` and `_fail_oog`. This means that the gas conditions for `_fail`, `_pass_oog`, and `_fail_oog` are **not** monotone, i.e. not of the form `VGas >= G` for some `G`.

I'm wondering if these specs could be simplified somehow by making the specs that end in `REVERT` or run out of gas have the same form. Then we could potentially return to the simple world where `VGas >= G` is the necessary and sufficient condition for not failing due to insufficient gas, so we could get away with only having two specs: `_pass` and `_fail` again?

Thoughts @MrChico @mhhf ?

**Comment by livnev** [comment]
For the record, I tested the current state of this PR thoroughly with the following (needlessly complicated) workflow:

```
klab run --spec out/specs/Callee_add_pass.k
klab solve-gas
klab build --exact-gas
klab run --spec out/specs/Callee_add_pass.k

klab run --spec out/specs/Callee_add_fail.k
klab solve-gas
klab build --exact-gas
klab run --spec out/specs/Callee_add_fail.k

klab run --spec out/specs/Callee_add_pass_oog.k
klab run --spec out/specs/Callee_add_fail_oog.k

# needed add_fail for some reason
klab run --spec out/specs/Callee_tempDelta_pass.k

klab solve-gas
klab build --exact-gas

klab run --spec out/specs/Callee_tempDelta_pass.k
klab run --spec out/specs/Callee_tempDelta_pass_oog.k

klab run --spec out/specs/easyNest_calling_pass.k
klab solve-gas
klab build --exact-gas
# FAIL!
klab run --spec out/specs/easyNest_calling_pass.k

passes with this gas expression:
<gas> VGas => ( ( ( ( VGas -Int 1257 ) -Int ( 700 +Int ( ( VGas -Int 1957 ) -Int ( ( VGas -Int 1957 ) /Int 64 ) ) ) ) +Int `#if_#then_#else_#fi_K-EQUAL`( notBool( ( ABI_x ==Int 0 )  ),( ( ( VGas -Int 1957 ) -Int ( ( VGas -Int 1957 ) /Int 64 ) ) -Int 20491 ),( ( ( VGas -Int 1957 ) -Int ( ( VGas -Int 1957 ) /Int 64 ) ) -Int 5491 ) ) ) -Int 42 )
</gas>

klab run --spec out/specs/easyNest_calling_fail.k
klab solve-gas
klab build --exact-gas
klab run --spec out/specs/easyNest_calling_fail.k
```

Everything worked except for the two points mentioned above:
(a) `_fail` spec required when it isn't logically needed
(b) incorrect final gas expression inserted into spec

File: libexec/klab-compile (hunk 1 above)
Comment (mhhf) [reply libexec/klab-compile]: done

**Comment by mhhf** [comment]
> @mhhf Looking at the `multipleCalls` example, it seems like the file `Callee_tempDelta_pass.k` contains specs for both `Callee_add_pass` and `Callee_add_fail`. I wonder why it has included the reachability claim for the latter? It seems logically unrelated. And as a result `klab run` complains that `add_fail` needs to be proved first, which I don't think is the case.

I talked to @MrChico about this - for _pass its indeed the case that only pass needs to be present - for fail, both have to be present (pass + fail). So making both a dependency for both seems cleaner to me.

**Comment by livnev** [comment]
@mhhf I'd prefer if the dependencies were the logical dependencies and nothing more. In some cases you might be blocked on proving the fail behaviour but want to press on with proving the pass behaviour.

File: libexec/klab-report [hunk 3]
```diff
@@ -39,7 +39,7 @@ const badge = (name, status) => `<div class="badge">
       ${name}
     </span>
   </div>
-  <div class="status ${status == "SUCCESS" ? "green" : (status == "???" ? "yellow" : "red") }">
+  <div class="status ${status == "pass" ? "green" : (status == "???" ? "yellow" : "red") }">
```
Comment (MrChico) [inline libexec/klab-report]: Also here should be `ACCEPT`

**Comment by MrChico** [comment]
@livnev >There's another outstanding issue
Yes, although this is the same issue I mentioned. And it should be a fairly easy fix. Your haskell tool has a tokenizer, right? We can just make `klab solve-gas` output both the gas condition (as it already does) and just a tokenized version of the `raw.kast` final gas expr, which we insert in the `<gas>` cell.

**Comment by MrChico** [comment]
@livnev 
> I'm wondering if these specs could be simplified somehow by making the specs that end in REVERT or run out of gas have the same form. Then we could potentially return to the simple world where VGas >= G is the necessary and sufficient condition for not failing due to insufficient gas, so we could get away with only having two specs: _pass and _fail again?

This would definitely be the easiest solution. I realized I was doing this the wrong way when I tried doing this before. I think the correct way uses ensures. With `<statusCode> _ => FAILURE </statusCode>`, and the clause
```
ensures
FAILURE =/=K EVMC_SUCCESS
```
this catches not only oog failures, but also stack overflows and so on. 

We still need this double run to catch oog failures though, so even though we will only be dealing with `pass` and `fail`, they will only be exhaustive if gas expressions have been inserted (and idk, maybe one still wants to distinguish (as in keep as two different specs) the `pass-exact-gas` vs. `pass` case, as well as the `fail` vs. `fail-exact-gas` case?)

**Comment by livnev** [comment]
@MrChico That sounds good. The approach you were trying before, was that `<statusCode> _ => FAILURE </statusCode>` and `FAILURE =/=K EVM_SUCCESS` in the `requires`? I'm not sure why that wouldn't work?

Agreed that we need to keep the two-pass model, and I like the idea of having separate spec files for `_pass` and `_pass_exactgas` since then the workflow is more explicit somehow and doesn't involve overwriting specs. :+1:

**Comment by ehildenb** [comment]
You could also say something like `_ => _:ExceptionalStatusCode` instead of `_ => EVMC_SPECIFIC_STATUS_CODE`

**Comment by mhhf** [comment]
@ehildenb unfortunately ExceptionalStatusCode [doesn't include](https://github.com/kframework/evm-semantics/blob/52118d9827ff757674f5e52ddb4f9b4e37bf75b7/network.md#ending-codes) `EVMC_REVERT`:

**Comment by ehildenb** [comment]
Hmmm, we can make a subsort `FailureStatusCode` where `EVMC_REVERT` is added, and then subsort that into `EndStatusCode`? Would that make things simpler?

**Comment by MrChico** [comment]
I'm happy with the `ensures` solution. I haven't had an amazing experience dealing with K's sort checkers.

**Comment by MrChico** [comment]
Hmm, I actually baking `fail` and `oog` together is not as clean as one might like it to be. Joining `_succ_oog` with `_fail_oog` is not very nice as they have different `requires`. If we really wanted to bake them together we could have a big disjunction between the two, but I'm not sure K would handle that very well. Better I think would to be to go with 3 cases: `pass`, `pass_oog` and `fail`. @mhhf @livnev

File: libexec/klab-compile [hunk 4]
```diff
@@ -114,7 +111,7 @@ const behaviour = genBehaviour({
       });
 // console.log(JSON.stringify(behaviour, false, 2));
 boot.status = behaviour.success
-  ? "ACCEPT"
+  ? "SUCCESS"
```
Comment (MrChico) [inline libexec/klab-compile]: ?

Comment (mhhf) [reply libexec/klab-compile]: ¯\_(ツ)_/¯
must be a merge/rebase fuckup

**Comment by livnev** [comment]
I pushed the tidying changes from https://github.com/dapphub/klab/pull/100 onto this branch. `klab run` is now deprecated, `klab prove` is a wrapper around `kprove` with an optional `--dump` flag.

File: README.md [hunk 5]
```diff
@@ -114,10 +123,11 @@ klab build
 This will generate a fail and success reachability rule for each `act` of your specification in the `SafeAdd/out/specs` directory.
 
 ### Running proofs
-To explore a proof with the interactive klab GUI, use `klab run`, specifying which proof to explore using the `--spec` flag:
+
+To explore a proof with the interactive klab GUI, use `klab debug`, specifying which proof to explore using the `--spec` flag:
```
Comment (MrChico) [inline README.md]: `--spec` flag deprecated

File: README.md [hunk 6]
```diff
@@ -114,10 +123,11 @@ klab build
...
 
 ```sh
-klab run --spec out/specs/proof-SafeAdd_add_succ.k
+klab debug out/specs/proof-SafeAdd_add_succ.k
```
Comment (MrChico) [inline README.md]: Proof name should be `out/specs/SafeAdd_add_pass.k

**Comment by MrChico** [comment]
This makes `klab prove --dump` basically the same as `klab debug --headless`, right? Is there a reason to keep them both?

File: acts.md [hunk 7]
```diff
@@ -12,7 +12,7 @@ rule A => C
     requires B
     ensures D
 ```
-However, when `A` and `C` both express configurations describing the entire state relevant to an ethereum contract, such an expression can get quite [large](https://github.com/dapphub/k-dss/blob/6d536c91a407fd3ead44b9347e25b6790663a10a/out/specs/proof-Vat_tune_succ.k).
+However, when `A` and `C` both express configurations describing the entire state relevant to an ethereum contract, such an expression can get quite [large](https://github.com/dapphub/k-dss/blob/6d536c91a407fd3ead44b9347e25b6790663a10a/out/specs/Vat_tune_succ.k).
```
Comment (MrChico) [inline acts.md]: breaks the link

File: acts.md [hunk 8]
```diff
@@ -45,7 +45,7 @@ returns X + Y
...
 
-To verify these claims, we need access to the bytecode generated by the code above. You can generate the proof claims by running `klab build` in the `klab/examples/SafeAdd` directory. This fetches the bytecode from the `sol.json` file specified in `klab/examples/config.json`, and generates the K reachability rules (or proof claims) in the `klab/examples/out/specs` directory. You can run the proof by running `klab run --spec <path-to-reachability-rule.k>`. 
+To verify these claims, we need access to the bytecode generated by the code above. You can generate the proof claims by running `klab build` in the `klab/examples/SafeAdd` directory. This fetches the bytecode from the `sol.json` file specified in `klab/examples/config.json`, and generates the K reachability rules (or proof claims) in the `klab/examples/out/specs` directory. You can run the proof by running `klab run <path-to-reachability-rule.k>`.
```
Comment (MrChico) [inline acts.md]: run -> debug

**Comment by livnev** [comment]
@mhhf Can you change it so that the spec names in `obligations.batch` are just the basename, without the `.k` extension, please?

**Pull request merged**